IIPs/iip-0.md

```diff
@@ -0,0 +1,107 @@
+
+## Specification
+
+Each IPNN node (implementation) needs to consider following this rules.
+
+### Nostr Improvement Possibilities
+
+At the first step, each IPNN nodes is a Nostr relay. so, IPNN nodes **MUST** have [NIP-01](https://github.com/nostr-protocol/nips/blob/master/01.md) implemented.
+
+> NOTE: the Nostr protocol itself doesn't focused on saving and maintaining events and it's more about communication and transmission of events. but for IPNN nodes, it's **RECOMMENDED** to save and maintain events they receive from clients or other nodes. limiting the space, backup, removing old events or some specific events and more, is **OPTIONAL**.
+
+### Network
+
+All IPNN nodes **MUST** support `tcp` and `udp` transports. supporting `quic` is **RECOMMENDED**.
+
+Port `37771`  for network is **RECOMMENDED**.
+
+### IPNN P2P protocol
+
+The IPNN P2P protocol is based on [libp2p protocols and specifications](ipns://docs.libp2p.io/concepts/fundamentals/protocols/).
+
+#### Topics
+
+There are 2 topics in IPNN base protocol:
+
+```
+/ipnn/topic/events/v1
+```
+
+The `events` topics is for broadcasting new events that each node gets from clients. every one **SHOULD** publish and **MUST** subscribe to this topic. All data in this topic **MUST** be an IPNN message.
+
+```
+/ipnn/topic/profiles/v1
+```
+
+The `profiles` topics is for broadcasting new profile and users a relay finds. each node **SHOULD** gossip new profiles in profiles topic. All data in this topic **MUST** be an IPNN message.
+
+#### Protocols
+
+There is a stream protocol in IPNN like this:
+
+```
+/ipnn/stream/v1
+```
+
+Every node can make a stream connection with other nodes and send and receive data. they **MUST** be in IPNN messages style.
+
+##### Messages
+
+The message in IPNN is defined by IPNN not libp2p.
+
+There is 2 message types in IPNN:
+
+###### Event
+
+Event message will be broadcasted in event topic or sent as query message response. it **MUST** be a valid Nostr event in *json*.
+
+This message is a event message when a some event is not found:
+
+```json
+{
+    "error":"not_found"
+}
+```
+
+###### Profile
+
+Profile message **MUST** contain a Nostr profile details in *json*.
+
+This message is a profile message when a profile is not found:
+
+```json
+{
+    "error":"not_found"
+}
+```
+
+###### Query
+
+Query message **MUST** be a json message like that:
+
+```json
+{
+    "type" : "profile" | "event",
+    "id" | "public_key" : "..."
+}
+```
+
+Receivier node **SHOULD** send a event message or profile message as answer.
```

Comment:
As per comment before, a receiver node answer according to the _kind_ being asked.